backend/chainlit/callbacks.py

```diff
@@ -87,6 +91,38 @@ async def oauth_callback(provider_id: str, token: str, raw_user_data: Dict[str,
     return func
 
 
+@trace
+def custom_authenticate_user(func: Callable[[str], Awaitable[User]]) -> Callable:
+    """
+    A decorator to authenticate the user via custom token validation.
+
+    Args:
+        func (Callable[[str], Awaitable[User]]): The authentication callback to execute.
+
+    Returns:
+        Callable[[str], Awaitable[User]]: The decorated authentication callback.
+    """
+
+    if len(get_configured_oauth_providers()) == 0:
+        raise ValueError(
+            "You must set the environment variable for at least one oauth provider to use oauth authentication."
+        )
+
+    config.code.custom_authenticate_user = wrap_user_function(func)
+    return func
+
+
+def custom_oauth_provider(func: Callable[[], OAuthProvider]) -> None:
```

Comment:
I am not sure whether `custom_` is the right name here. As we'll be moving to having less and less oauth providers built in, I feel it's more likely that _most_ if not _all_ oauth providers will be pluggable. Additionally, I want to move towards using an externally maintained OAuth client framework instead (see #1240). 

So perhaps, this could become the _default_ way of setting up oauth providers -- in that case something like `@register_oauth_provider` or merely `@oauth_provider` might be a more appropriate name.